pkg/agent/agent.go

```diff
@@ -136,16 +151,34 @@ func (k *Klocksmith) process(stop <-chan struct{}) error {
 		return err
 	}
 
+	glog.Info("Checking if node is already unschedulable")
+	node, err = k8sutil.GetNodeRetry(k.nc, k.node)
+	if err != nil {
+		return err
+	}
+	alreadyUnschedulable := node.Spec.Unschedulable
+
 	// drain self equates to:
-	// 1. set Unschedulable
+	// 1. set Unschedulable if necessary
 	// 2. delete all pods
 	// unlike `kubectl drain`, we do not care about emptyDir or orphan pods
 	// ('any pods that are neither mirror pods nor managed by
 	// ReplicationController, ReplicaSet, DaemonSet or Job')
 
-	glog.Info("Marking node as unschedulable")
-	if err := k8sutil.Unschedulable(k.nc, k.node, true); err != nil {
-		return err
+	if !alreadyUnschedulable {
+		glog.Info("Marking node as unschedulable")
+		if err := k8sutil.Unschedulable(k.nc, k.node, true); err != nil {
```

Comment:
I believe there's a slight race here.

If the node is marked unscheduleable, but the annotation call fails (e.g. due to apiserver issues or this code failing), the failure condition is that the node will be unschedulable and never be made schedulable again.

I think that the `k8sutil.SetNodeAnnotations` call several lines above should include this annotation if necessary.

I think that setting the annotation before marking it unschedulable leaves it in a state where it will ultimately do the right thing once it succeeds.

Reply:
Ah, right. Good catch. Thanks!

Reply:
Looks like this was done at line [165](https://github.com/coreos/container-linux-update-operator/pull/176/files#diff-64c23753751bef052315a093bb62faeaR165).

Reply:
I believe this got updated after this comment and github just didn't mark it as resolved